lxd/auth/types.go

```diff
@@ -52,6 +52,12 @@ type Authorizer interface {
 	// the entity. The effective project for the entity must be set in the request context as request.CtxEffectiveProjectName
 	// *before* the call to GetPermissionChecker.
 	GetPermissionChecker(ctx context.Context, entitlement Entitlement, entityType entity.Type) (PermissionChecker, error)
+
+	// AddEntitlements attaches entitlements to an entity.
+	AddEntitlements(ctx context.Context, entityType entity.Type, entityURL *api.URL, entity entity.EntityWithEntitlements) error
+
+	// AddEntitlementsToEntities attaches entitlements to multiple entities.
+	AddEntitlementsToEntities(ctx context.Context, entityType entity.Type, entityURLs []*api.URL, entities []entity.EntityWithEntitlements) error
```

Comment:
I'm not sure on these function signatures. Especially in the second case where the caller must prepare a list of URLs to correspond by index to the list of entities. 

It seems odd to me that we're passing in e.g. an `api.Instance`, along with it's URL and it's type, both of which can be derived from the fact that it is an `api.Instance`.

Do you think it's worth adding two more methods to `EntityWithEntitlements` e.g.:
```
type EntityWithEntitlements interface {
    AddEntitlement(auth.Entitlement)
    URL() string
    EntityType() entity.Type
}
```

Then these function signatures can become:
```
AddEntityEntitlements(ctx context.Context, entity entity.EntityWithEntitlements)
AddEntitiesEntitlements(ctx context.Context, entities entity.EntityWithEntitlements)
```

Let me know what you think.

Reply:
I like the the idea but I'd prefer to call the `URL() string` method `EntityURL() string` instead to be more explicit. For example, the URL scheme used for an LXD instance can be defined like:
`url := api.NewURL().Path(version.APIVersion, instancePath, instance.Name).Project(instance.Project)`
and if `instance.Project` is `default` then no query param will be added whereas an OpenFGA entity URL will systematically have the `project` query param set even if its `default`.

Reply:
@markylaing Actually, I realize this will be a bit more complicated than that:

There is an existing logic for URL generation in the `shared/entity` package, but I won't be able to use that from the `shared/api` package (circular dependency issues) where my implementations of `EntityWithEntitlements` are defined...

Reply:
@markylaing can't we just move the `url.go` file from `shared/entity` into the `shared/api` package? We'd keep the `type.go` in `shared/entity`. This way we can reuse the existing `<Entity>URL` functions for the API LXD entity that would implement  the `EntityWithEntitlements` interface. I think this would make the circular dep issues go away..

Reply:
This seems difficult. Instead could we do something like:

```go
type EntityWithEntitlements interface {
	AddEntitlement(entitlement string)
	EntityType() Type
}

type EntityWithEntitlementsAndURL struct {
	Entity EntityWithEntitlements
	EntityURL    *api.URL
}

...

// In the interface
AddEntityEntitlements(ctx context.Context, entity entity.EntityWithEntitlementsAndURL)
AddEntitiesEntitlements(ctx context.Context, entities []*entity.EntityWithEntitlementsAndURL)
```

This way, we can set the EntityURL in the `lxd` package when calling the `AddEntityEntitlements` / `AddEntitiesEntitlements` function without having any issues. What do you think?

Reply:
Can we just define the interface somewhere else? In fact, it is only used internally, so perhaps it is better placed in `lxd/auth`? Would that help with the circular dependencies?